### ISSUE TITLE
NumPy 2 bringup

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/conda-forge.github.io#1997]
NumPy is currently working on [a NumPy 2.0 release]( https://github.com/orgs/numpy/projects/9 ), which is planned to come out later this year. Here are [the current (draft) release notes]( https://numpy.org/devdocs/release/2.0.0-notes.html ). Also here's the upstream tracking issue ( https://github.com/numpy/numpy/issues/24300 ), and ecosystem compatibility [tracker](https://github.com/numpy/numpy/issues/26191).

Some questions worth discussing:

* How do we handle tightening loose `numpy` pins in packages ( https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/issues/516 )?
* Do we want to keep building for NumPy 1 & 2 at the same time?
* What timeline do we want to use for adding NumPy 2?
* What timeline do we want to use for dropping NumPy 1?
* How should this be coordinated around other major migrations (like Python 3.12 or a new Boost)?
* Numpy now builds for the "oldest-available-C-API" by default, i.e. we could build for our current 1.22 C-API using numpy 1.25; we need to adapt to this in any case, at the latest when our lower bound becomes 1.25 (see https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/4816)?
* Anything else we should discuss?

Todos:

* [x] https://github.com/conda-forge/numpy-feedstock/pull/313
* [x] https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/728
* [x] https://github.com/regro/cf-scripts/issues/2469
* [x] https://github.com/regro/cf-scripts/pull/2470
* [x] https://github.com/conda-forge/conda-smithy/issues/1911
* [x] https://github.com/conda-forge/staged-recipes/pull/26188
* [x] https://github.com/conda-forge/numpy-feedstock/pull/314
* [x] https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5790
* [x] https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5851
* [x] https://github.com/conda-forge/conda-forge.github.io/issues/2156

cc @conda-forge/core

**Comment by h-vetinari** [comment]
See also https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/4816

**Comment by ocefpaf** [comment]
Can we merge these two issues just to make it easier to track them?

**Comment by jakirkham** [comment]
The issue Axel raises seems like a subpoint of this issue (depending on what we decide). Namely do we want to opt-in to this newer/slimmer ABI and how does that fit into NumPy 2

**Comment by ocefpaf** [comment]
Sure. IMO Axel issue is a subset of this one. I don't have strong opinions on which one to keep, or if you want to keep both, but I also don't want to get lost on two mega-threads :grimacing:

**Comment by jakirkham** [comment]
Added Axel's item to the list above

**Comment by h-vetinari** [comment]
Handling the ABI is the key point here (that and current packages missing a `<2`). I updated the added item because the summary was not accurate.

Normally I'd say we do a dual migration (keep 1.x; add 2.0), but numpy has around 5000** dependents in conda-forge, so that would be a pretty substantial CI impact, especially if it takes a while to drop 1.x.

**
```
>mamba repoquery whoneeds numpy -c conda-forge -p linux-64 > tmp
># edit to remove header
>python
>>> q = open("tmp", "r").readlines()
>>> p = {x.strip().split(" ")[0] for x in q} - {""}
>>> len(p)
4898
```
Obviously not all of them are compiling against numpy, but still...

**Comment by jakirkham** [comment]
> I updated the added item because the summary was not accurate.

Thanks Axel! 🙏 

Anyone should feel free to update the issue as needed 🙂

**Comment by jakirkham** [comment]
Following up on our discussion earlier about improving the visibility of `NPY_FEATURE_VERSION`, started this NumPy PR ( https://github.com/numpy/numpy/pull/24861 ) to message how the value is set

Also include a note about one approach we might take to ensure that value is embedded in built binaries. Though maybe there are better approaches for that portion

**Comment by jakirkham** [comment]
There should now be a string baked into binaries built with NumPy to notate what kind of NumPy compatibility they have

https://github.com/numpy/numpy/pull/25948

**Comment by jakirkham** [comment]
It is worth noting that thanks to Axel and others we now have NumPy 2.0.0rc1 packages: https://github.com/conda-forge/numpy-feedstock/issues/311

Also ecosystem support of NumPy 2 is being tracked in this issue Ralf opened: https://github.com/numpy/numpy/issues/26191

We are now in a good spot to start testing building packages with NumPy 2

**Comment by h-vetinari** [comment]
I discussed this with @rgommers recently and one important point that he brought up is the situation with `pin_compatible`, which we'll have to fix as part of any migration effort, probably with a piggyback migrator, since we'll need to rewrite the recipes.

In particular, since `numpy` isn't separated into a library and run-time component, we don't have a run-export, and so feedstocks use `pin_comparible` under `run:`. However this will be doubly incorrect in the new setup - for one as our `NPY_FEATURE_VERSION` (which forms the lower bound) will be _lower_ than the one at build time, and second because the upper bound should be something like `<2.{{ numpy.split(".")[1] | int + 3 }}` (for a project that's free of deprecation warnings; anything else might be deprecated in `2.{N + 1}` and removed after two releases in `2.{N + 3}`).

**Comment by h-vetinari** [comment]
> In particular, since `numpy` isn't separated into a library and run-time component, we don't have a run-export [...]

Of course, if there's appetite for a split into `libnumpy` (with a run-export) and `numpy` (the python bits), that might be worth a thought as well. But then even moreso, we'd need a piggyback.

**Comment by isuruf** [comment]
> In particular, since numpy isn't separated into a library and run-time component, we don't have a run-export, and so feedstocks use pin_compatible under run:

We do have a run_export on numpy.

**Comment by h-vetinari** [comment]
Yeah... Clearly I shouldn't be writing these comments from a phone 😅

I misremembered that part, but in that case the task becomes easier - we just set up the right run export in numpy itself, and then remove `pin_compatible` in feedstocks that compile against numpy. Right?

**Comment by h-vetinari** [comment]
Another question we have to answer soon: what mechanism do we want to use for setting `NPY_FEATURE_VERSION`... Perhaps the easiest would be an activation script in `numpy`, but that's a fairly big hammer, as it persists beyond build time and into all user environments.

Right now I'm thinking of setting `NPY_FEATURE_VERSION` in the global pinning (cleanly overrideable per feedstock where necessary), and then using that in conda-forge-ci-setup to populate the environment variable that `numpy` will pick up (and if necessary, in the compiler activation feedstocks, e.g. for `CFLAGS`).

The only issue there is that the run-export on `numpy` is not dynamic, in the sense that it gets fixed to the value of `NPY_FEATURE_VERSION` at the build time of numpy, and not the (potentially different) one in play when building something else against numpy. Unless I'm overlooking something, we'd therefore need to _transform_ rather than remove the existing uses of `pin_compatible("numpy")` with something like
```
- {{ pin_compatible("numpy", lower_bound=NPY_FEATURE_VERSION) }}
```
while the upper bound (`<2.{N + 3}`) would be set by the run-export on `numpy`.

**Comment by jakirkham** [comment]
What if we had something like...?

```yaml
{% set version = "2.0.0" %}

package:
  name: numpy
  version: {{ version }}

...

build:
  ...
  run_exports:
    - {{ pin_subpackage("numpy", lower_bound=os.environ.get("NPY_FEATURE_VERSION", version)) }}
```

That way we can defer this environment variable setting to packages

If they don't set something, we can provide a sensible default (either `version` or something else we decide)

We could also consider whether conda-build could allow `NPY_FEATURE_VERSION` to be a pass through environment variable or if we handle that within conda-forge with some recipe changes to pass it through ourselves. This would let us set us use a global setting (as you suggest)

**Comment by rgommers** [comment]
I don't think this type of `NPY_FEATURE_VERSION` is useful at all. NumPy guarantees to set it to a version that is not higher than the first numpy release that supported the Python minor version being built for. So all produced extension modules will work with all possible numpy versions that can actually be installed.

Hence, doing nothing should be the right default here, trying to change it from NumPy's default will likely only be the cause of extra complexity/confusion, and perhaps bugs.

**Comment by h-vetinari** [comment]
> That way we can defer this environment variable setting to packages

I'd be surprised if it works like that. AFAIU, that `os.environ` call will be resolved while building numpy.

> I don't think this type of `NPY_FEATURE_VERSION` setting is useful at all. NumPy guarantees to set it to a version that is not higher than the first numpy release that supported the Python minor version being built for.

Leaving aside NEP29, this is a quantity we have to be able to control IMO. Otherwise our metadata for packages building against numpy is bound to be wrong, and deteriorate over time (when numpy inevitably moves the lower bound, and we don't move in sync across all our feedstocks). I don't see how we can reasonably avoid making `NPY_FEATURE_VERSION` explicit in conda-forge in some way.

**Comment by jakirkham** [comment]
I very well could be wrong. It is easy to test

Really we just need more ideas to sample from. It's more important that we have a large variety before selecting one. So feel free to propose more

**Comment by rgommers** [comment]
> Otherwise our metadata for packages building against numpy is bound to be wrong, and deteriorate over time (when numpy inevitably moves the lower bound, and we don't move in sync across all our feedstocks).

It won't be wrong. The metadata that is in the upstream releases (i.e. the `dependencies` key in `pyproject.toml` files) is going to be updated by package authors, and that's the thing that should be relied on by conda-forge builds. The build-time version of numpy is now simply taken off the table completely, it no longer adds an extra constraint.

**Comment by h-vetinari** [comment]
I'm not sure I follow. Say a project has `numpy >=1.24,<2.3` in its `pyproject.toml`, is there some sort of hook that populates `NPY_FEATURE_VERSION` to 1.24? If so, how would that constraint arrive in the metadata of the packages we build?

Or do you mean that the default for that in numpy is so low (1.19?) that it won't ever become a factor? That seems doubtful to me.

Even aside from those questions, we still have an interest to provide a common baseline for numpy compatibility (so that most of conda-forge is still usable with the oldest supported numpy), and avoid that individual packages move on too quickly (unless they really need to), or extremely slowly (i.e. going back to 1.19 adds about 2 years on top on top of what NEP29 foresees w.r.t. being able to use a given ABI feature).

> The build-time version of numpy is now simply taken off the table completely, it no longer adds an extra constraint.

In summary, this seems highly dubious to me. There's still a lower bound _somewhere_, either in numpy's default feature level, or in an explicit override of `NPY_FEATURE_VERSION`. However it comes to be, we should represent that lower bound in our package metadata exactly (or at the very least, something tighter).

**Comment by rgommers** [comment]
> Or do you mean that the default for that in numpy is so low (1.19?) that it won't ever become a factor? That seems doubtful to me.

This. And it's not doubtful, it is guaranteed to work. The whole point is to take away build-time version as a thing that dynamically overrides the declares runtime dependency range. 

> Even aside from those questions, we still have an interest to provide a common baseline for numpy compatibility (so that most of conda-forge is still usable with the oldest supported numpy), and avoid that individual packages move on too quickly

No, that does not make sense. If a package has `numpy>=x.y` in its constraints, you cannot just ignore that. The package author bumped the lower version for some reason, so if you tweak the metadata to say numpy>=x.y-N` instead, you will allow a broken combination of packages.

> In summary, this seems highly dubious to me. There's still a lower bound somewhere, either in numpy's default feature level, or in an explicit override of NPY_FEATURE_VERSION. However it comes to be, we should represent that lower bound in our package metadata exactly (or at the very least, something tighter).

No, and no. The lower bound is whatever `dependencies=numpy...` in pyproject.toml` says, or it's a bug in the package (even if the package internally sets NPY_FEATURE_VERSION, which should be quite rare).

What the conda-forge tooling should do is check that the `environment.yml` and `pyproject.toml` metadata is consistent - and I think that is a feature already present for regular Python packages. I.e., start treating `numpy` like any other Python package when building against numpy 2.x.

**Comment by h-vetinari** [comment]
> No, that does not make sense.

You chopped off the part of my quote that accounts for the scenario you describe.

> The lower bound is whatever `dependencies=numpy...` in `pyproject.toml` says

I'm not saying we should disregard runtime constraints. I'm saying we also need to express constraints arising from the feature level - both of those can be attached to the same package without conflict. They stack and the intersection of both is what's actually permissible for the final artefact.

> What the conda-forge tooling should do is check that the `environment.yml` and `pyproject.toml` metadata is consistent

I don't see this happening soon enough to be available for the 2.0 transition, it would need work on conda-build AFAICT.

> and I think that is a feature already present for regular Python packages. I.e., start treating `numpy` like any other Python package when building against numpy 2.x.

I'm not sure what you mean here. Presumably by "python package" you don't mean "pure python" packages? Anything else that has a run-export (to my knowledge) uses the build-time version as a lower bound. That's precisely the issue that requires attention here, because of the very unusual situation how building against numpy 2.0 produces something compatible with `>=1.x`.

**Comment by rgommers** [comment]
> I'm saying we also need to express constraints arising from the feature level - both of those can be attached to the same package without conflict. They stack and the intersection of both is what's actually permissible for the final artefact.

What I am trying to explain is that that stacking is not doing anything, because
- `numpy` will never set the feature version in a way that allows for this to have any effect, and
- if a package does this internally by setting `NPY_FEATURE_VERSION` to something higher than what it says in its `pyproject.toml`, that's a bug in the package and should be fixed there by fixing its `dependencies=` metadata.

> I'm not sure what you mean here. Presumably by "python package" you don't mean "pure python" packages? 

Ah, I did mean this since I remember dependencies being flagged on PRs - but it may not be ready indeed, since it's marked as experimental:

<img width="323" alt="image" src="https://github.com/conda-forge/conda-forge.github.io/assets/98330/299ccb83-a075-4da3-a579-73c34bafa3c8">
<img width="584" alt="image" src="https://github.com/conda-forge/conda-forge.github.io/assets/98330/83a4694e-78c4-4720-bafd-3fbafa6752b3">

* * *

So it's still mostly manual then, depending on the feedstock maintainers to keep `pyproject.toml` and `environment.yml` in sync?

**Comment by rgommers** [comment]
> For clarification, should `environment.yml` here be the recipe's `meta.yaml`? Or do you mean something else Ralf?

Yes indeed. General tiredness 🤦🏼. Editing my comment to say `meta.yaml` to avoid further confusion.

**Comment by jakirkham** [comment]
Thanks Ralf! 🙏 All good. Appreciate hearing your insights and having your support here 🙂

Can imagine there are a lot of spinning plates with this work 😅

**Comment by rgommers** [comment]
> Anything built against numpy 2.0 (assuming I undestood correctly that the default ABI level is going to be 1.19) needs to be impossible to install with 1.18, hence needs the right constraints.

It is impossible. The lowest Python version supported by NumPy 2.0 is `3.9`. There is no `numpy` package for 1.18 on either conda-forge or on PyPI:
```
$ mamba search numpy=1.18.5
Loading channels: done
# Name                       Version           Build  Channel             
numpy                         1.18.5  py36h7314795_0  conda-forge         
numpy                         1.18.5  py36he0f5f23_0  conda-forge         
numpy                         1.18.5  py37h8960a57_0  conda-forge         
numpy                         1.18.5  py38h8854b6b_0  conda-forge
```

So yes, it is _by design of how numpy sets the default targeted C API_ impossible to get an incompatible combination here, even in an example like cvxpy where they set their lower bound to 1.15 (that could well be valid if they still support Python 3.7, no idea).

> For example, we're still building for py38

There cannot be a numpy 2.0 package for py38, so this isn't relevant.

> So we need to deal with this.

I'm still convinced that this is not true - it does not need dealing with explicitly in conda-forge recipes because it cannot go wrong. It's perfectly okay for the conda-forge `cvxpy` version to have `>=1.15` in its metadata, that will work for any actual `numpy` package built by conda-forge.

**Comment by h-vetinari** [comment]
> So yes, it is _by design of how numpy sets the default targeted C API_ impossible to get an incompatible combination here

OK, that is the key piece I was missing. The necessary `>=1.x` I was talking about is enforced implicitly, simply by the lack of availability of older numpy builds for whatever Python version we're using.

So, for my understanding (and walking through the logic), once numpy drops support for a given Python version, the default `NPY_FEATURE_VERSION` can move up (ignoring for now that not every release changes ABI) until the first numpy version that supported the currently oldest-supported Python version? Is that the idea?

**Comment by rgommers** [comment]
Yes, that sounds right. To rephrase that:
- numpy's `pyproject.toml` contains and enforces the lowest-supported version, e.g.: `requires-python = ">=3.9"`
- the first numpy version to support that python version (e.g., `1.19.0` for py39) determined the maximum version that `NPY_FEATURE_VERSION` may be set to
- when that `>=3.9` is bumped to `>=3.10`, then that max-allowed version for `NPY_FEATURE_VERSION` moves up.

We must be getting this right in numpy, or there'll be a ton of issues. And I don't want conda-forge to have to worry about setting `NPY_FEATURE_VERSION` for several reasons:
- It'd be extra boilerplate/churn in every recipe
- It'd make what is designed to be a rarely-used thing that we'd prefer downstream package authors to not have to know about now a prominent thing
- If it would be needed, then it'd be needed for other distros that support multiple package versions (e.g., Spack too) - and I want to be sure that that's not the case.

**Comment by jakirkham** [comment]
Thanks @rgommers , @h-vetinari , and @isuruf for joining the meeting today! 🙏

My understanding of next steps from our discussion are to

* Start a piggyback migrator to drop `{{ pin_compatible("numpy") }}` from recipes
  * Our `run_exports` handles this correctly for NumPy 1
  * With NumPy 2, `pin_compatible` will be unneeded as NumPy targets the oldest NumPy C API for a given Python version (say 3.9)
* Update the `run_exports` in our NumPy 2 packages to make NumPy 1.22 the minimum supported version
* Start a NumPy 2 migrator that uses our RC packages (similar to what was done for Python 3.12)
  * Maybe we should configure this to a very small number or low frequency of PRs to start so we can work out any kinks

Are there any other steps we should consider? Anything we should revise in the steps above?

Once we are happy with the list, will add a checklist to the OP of this issue to track

Please let me know what you think 🙂

**Comment by rgommers** [comment]
That sounds about right to me. Before that, https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/issues/516 may need checking/doing?

**Comment by h-vetinari** [comment]
> My understanding of next steps from our discussion are to
> * Start a NumPy 2 migrator that uses our RC packages (similar to what was done for Python 3.12)

This is the first step that's necessary, but that already runs into several problems, see https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5790

> * Start a piggyback migrator to drop `{{ pin_compatible("numpy") }}` from recipes

We need to write this (not so difficult) and attach it (in regro/cf-scripts) to the `numpy2` migrator from the previous point. I'll do this when we get closer to having a mergeable `numpy2` migrator.

> * Update the `run_exports` in our NumPy 2 packages to make NumPy 1.22 the minimum supported version

This one's easy: https://github.com/conda-forge/numpy-feedstock/pull/313

> * https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/issues/516) may need checking/doing?

Done: https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/712

**Comment by jakirkham** [comment]
Thanks Axel and Ralf! 🙏

Created a list in the OP linking PRs or issues. Please double check that. Happy to update as needed

For the piggyback migrator, think we don't need to wait actually. More details in issue ( https://github.com/regro/cf-scripts/issues/2469 ). Happy to discuss further there

**Comment by isuruf** [comment]
There's no need to build for both 1.x and 2.x at the same time. If a package does not support numpy 2, the migration is held at that point and we wait until that package support numpy 2.

**Comment by jakirkham** [comment]
Agreed the migrator can just replace NumPy 1.22 with NumPy 2.0. Both will produce packages that work on NumPy 1.22+ (the latter providing NumPy 2 support)

**Comment by h-vetinari** [comment]
> There's no need to build for both 1.x and 2.x at the same time.

Of course there is - pulling forward the building of the 2.0-compatible builds before the release of numpy 2.0 GA. This is exactly what we did for python 3.12.0rc. Once the GA release happens, we can then publish numpy 2.0 into main, and all the other packages built against 2.0rc1 will be installable in a numpy 2.0 world already.

**Comment by h-vetinari** [comment]
> pulling forward the building of the 2.0-compatible builds before the release of numpy 2.0 GA. This is exactly what we did for python 3.12.0rc.

This is what the consensus in the core call was. It may well turn out to be hard/impossible due to constraints around pinning/zips/smithy, but that was the desire at least.

**Comment by jakirkham** [comment]
Maybe I'm misunderstanding. Do we want to publish packages built using NumPy 2.0.0rc1 to a different label (say to do additional testing on them before releasing them in the wild)? Or do we just want to publish them to `main`?

**Comment by h-vetinari** [comment]
> Maybe I'm misunderstanding. Do we want to publish packages built using NumPy 2.0.0rc1 to a different label (say to do additional testing on them before releasing them in the wild)? Or do we just want to publish them to `main`?

So there's been a fair amount of relevant discussion in https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/712 that should probably be moved here, but for completeness:

What we did for python [3.12.0rc](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4914) was that we pulled `python` itself from a label (`conda-forge/label/python_rc`), but the packages built against 3.12 were published to `main`. Still, they couldn't be installed, because they depended on a python-build that was only available by explicitly opting in with `-c conda-forge/label/python_rc`. However, because of the ABI stability guarantee of CPython 3.12.0rc vs. GA, the moment we published `python 3.12.0` to `main`, all those builds became installable (we had roughly ~1000 feedstocks done by the GA release).

Numpy give the same guarantees about ABI stability between rc & GA, and so we could do exactly the same thing. I.e. packages built against numpy publish to `main`, but depend on a `numpy` that requires an opt-in channel. That would give us time to roll thing out in advance, but then the only way to roll that out over a period of ~weeks before GA release is if we do the 2.0 builds _additively_.

**Comment by seberg** [comment]
> packages built against numpy publish to `main`, but depend on a `numpy` that requires an opt-in channel

Except that there is no _runtime_ dependency on a newer NumPy: That package will be installable right away since it also works with old NumPy versions.
The logic you should need is basically:
* Use NumPy 1.22 for Python 3.8 (as newer version don't support 3.8, so they may not compile, even if the headers would probably be OK).
* Use the newest available NumPy version (right now 2.0) to compile for all other Python versions.
* It _may_ be nice to honor an additional runtime/buildtime pin, but TBH, I don't think it matters: Even if a package only supports 1.x at runtime as long as compilation succeeds it'll be fine to compile it with NumPy 2.  (I.e. it may be sensible to try with 2.0 and add the constraint if it fails.)

(Quite likely, I am misreading, but it looked a bit like that try wanted to build two versions, which would be fine but seems unnecessarily complicated.)

**Comment by h-vetinari** [comment]
> Except that there is no _runtime_ dependency on a newer NumPy: That package will be installable right away since it also works with old NumPy versions.

You're right about that, the lower run-time dependence changes the calculus (unless we add another constraint on packages built during the RC phase, which I would have ended up with, see below).

> Use NumPy 1.22 for Python 3.8

That's a good point actually; I think we should seriously consider dropping 3.8, rather than complicating the numpy 2.0 setup into a before/after 3.9 thing.

> Even if a package only supports 1.x at runtime as long as compilation succeeds it'll be fine to compile it with NumPy 2.

My assumption was we don't yet want to build against the RC to immediately be generally available, but if people are fine with that, why not?

**Comment by seberg** [comment]
> but if people are fine with that, why not?

I would say this is a gamble that may be fine (assuming the build succeeds).  It seems pretty unlikly to go bad and if it fails would also seem to point to critical NumPy issues?

Basically, it isn't even a gamble: SciPy, matplotlib, pandas, etc. etc. already do this for wheels!

The one issue I can think of is weirder constructs that we saw in dependencies of `eigenpy`: `eigenpy` depedenecies may use NumPy _through_ `eigenpy` which effectively makes them a single compilation unit from NumPy's perspective.  That could be problematic unless they are transitioned all in one go (pinning each others versions).  Likely, that isn't even a problem for conda, though.

**Comment by jakirkham** [comment]
> My assumption was we don't yet want builds against the RC to immediately be generally available, but if people are fine with that, why not?

Either approach seems potentially reasonable

We could start by putting a few packages built with the NumPy in a special label and test them out for a bit. Then, once we are comfortable, flip over to building and publishing them to `main`

Had thought maybe we want to do this with a few core ones like SciPy, Matplotlib, Pandas, some of the scikit-*s. That said, no strong feelings on this approach

**Comment by h-vetinari** [comment]
I updated https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5790 to build against 2.0.0rc1 directly; it looks like it still needs a smithy fix, see https://github.com/conda-forge/conda-smithy/issues/1911

**Comment by h-vetinari** [comment]
Migration is a-go! :)

**Comment by h-vetinari** [comment]
Oh, you know, just a little migration... \*checks [notes](https://conda-forge.org/status/migration/numpy2)\*... 2800 affected feedstocks 🫣😅

**Comment by h-vetinari** [comment]
OK, it looks like that number is not correct - one of the packages that stood out to me was [importlib_metadata](https://github.com/conda-forge/importlib_metadata-feedstock/blob/main/recipe/meta.yaml), which doesn't depend on numpy at all. It's probably not a great sign that this is being picked up.

**Comment by hmaarrfk** [comment]
Just so it doesn't get lost in the notifications, something seems to be off with the python 3.12 pinning. I know it is confusing to do dual migrations at the same time so I'm not sure there is an easy answer
https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5790#discussion_r1590007898

**Comment by h-vetinari** [comment]
Yeah, thanks for the note @hmaarrfk & @xhochy. In previous numpy bumps, we always did them for the open python 3.x & pypy migrations as well, and this wasn't an issue. Here however, bumping the 3.12 migrator gets immediately picked up by a given feedstock rerendering and so runs into breakage if that package isn't yet numpy 2.0-compatible (basically all of them).

I undid the changes to the 312 migrator to unbreak this, but I'm kind of unsure how to square this circle now. I think for now the least bad option is to just accept that we don't get numpy 2.0 builds for 3.12. Alternatives might be:
1. trying to override the whole numpy/python/python_impl zip in the numpy2 migrator to include 3.12 directly.
    * I tested this and it seems to work
    * Downside there is that this will open PRs to all python-related feedstocks, because the migrator bot sees `python:` as being updated.
2. Close Python 3.12 migration (which would be waaaaay earlier than we usually do this)

With 1., the numpy2 migrator would look like
```
numpy:
  - 1.22  # no py38 support for numpy 2.0
  - 2.0
  - 2.0
  - 2.0
  - 2.0
python:
  - 3.8.* *_cpython
  - 3.9.* *_cpython
  - 3.10.* *_cpython
  - 3.11.* *_cpython
  - 3.12.* *_cpython
python_impl:
  - cpython
  - cpython
  - cpython
  - cpython
  - cpython
```
while the python312 one would stay untouched.

**Comment by h-vetinari** [comment]
Next [snag](https://github.com/regro/cf-scripts/issues/2532): bot isn't opening PRs 😑

**Comment by jakirkham** [comment]
We are 87% of the easy through the Python 3.12 migration. Maybe we are close to closing this out? Perhaps with a little effort?

Wonder if there are other configuration options available to us if we make the NumPy 2 migration depend on the Python 3.12 migration? IOW the "no ice cream until you eat your vegetables" approach 😉

**Comment by h-vetinari** [comment]
> Wonder if there are other configuration options available to us if we make the NumPy 2 migration depend on the Python 3.12 migration?

Yeah, there's `wait_for_migrators`, but this is not exactly well tested and it would still run into the issue that we'd have to override the whole zip, which would then open PRs for all non-noarch python feedstocks.

**Comment by h-vetinari** [comment]
> We are 87% of the easy through the Python 3.12 migration.

The migrator status page is somewhat misleading on this - it reports the percentage of "PRs opened", not "PRs closed". So we're actually at ~67% with the 3.12 migration, with about ~900 feedstocks still to go. That's IMO quite a bit too early to close?

I guess the only nice alternative to points 1. & 2. [above](https://github.com/conda-forge/conda-forge.github.io/issues/1997#issuecomment-2094361188) would be to implement something in cf-scripts like `skip_unless_present:`, where we could say that a feedstock should only be migrated if it depends on `numpy`, in order to force it to ignore the feedstocks that wrongly get picked up due to the presence of the `python:` key (that we need to override the zip) if we do point 1.

I guess there aren't really any other zips that are affected in this way, so it'd be a bit tailor-made for the numpy/python situation. From that POV, maybe we could also just deal with the hassle of closing the superfluous PRs this one time, because once we're on numpy 2.0 (and have dropped 3.8 around October), we can completely remove numpy from the python zip[^1] going forward.

[^1]: except PyPy, but that's a separate discussion for various reasons.

**Comment by h-vetinari** [comment]
FWIW, I think we need to migrate 3.12 as well, as we won't simply be able to switch this one later (due to lots of missing dependencies then).

So if we don't do it as part of the overall numpy 2.0 migration, we'd need a separate migrator (now or later) that catches up 3.12 specifically. I feel that a second such migration wastes more time than it would do to deal with the overzealous PRs to python-but-not-numpy-related feedstocks (which by themselves aren't harmful even if merged).

**Comment by h-vetinari** [comment]
I've [paused](https://github.com/conda-forge/conda-forge-pinning-feedstock/commit/9f4bd780ca403d4862609c503c88b7ad27d6597b) the numpy2 migration for now until we decide how we want to handle this (as all PRs that are merged now would have to be cleaned up again later).

I've also opened a PR [proposing](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5851) to include python 3.12 in the current migrator.

**Comment by jakirkham** [comment]
After some discussion, we decided to soft-close the Python 3.12 migration. Meaning that the Python 3.12 migration is still running, but we now also include Python 3.12 in the global pinning. As a result, re-rendering a feedstock could add Python 3.12 (even if its dependencies are not available on Python 3.12 yet). Meaning the feedstock would get Python 3.12 builds that would fail. These could simply be ignored for now. Based on the Python 3.12 migration status, this is only relevant for ~11% of feedstocks (admittedly that is ~500 feedstocks). The rest have either already been migrated or have a migration PR

With Python 3.12 now addressed, we restarted the NumPy 2 migration today ( https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/5851 ). Here is the [status of NumPy 2 migration]( https://conda-forge.org/status/migration/numpy2 ). It may take it a bit to start adding PRs to all the feedstocks that can now take a PR. So please be patient 🙂

Thanks all! 🙏

**Comment by bashtage** [comment]
statsmodels still seeling some NumPy 1.22 builds on my NumPy 2 migration.  Have these not been migrated yet?  Or will they never be migrated (and so need to be skipped (and if so, any pointers would be helpful)).

![image](https://github.com/conda-forge/conda-forge.github.io/assets/5585221/745a704e-9549-43da-81ad-04c0301dd902)

These all fail.  All of the NumPy 2 targeted build pass (as do a few NumPy 1.22, although most of these should be disabled when the recipe is rerendered). 

[statsmodels NumPY 2 PR](https://github.com/conda-forge/statsmodels-feedstock/pull/67)

**Comment by h-vetinari** [comment]
> These all fail.

Those are all the PyPy builds, which is not participating in the numpy 2.0 migration. I haven't opened an issue yet, but PyPy support is (unfortunately!) being wound down, so don't hesitate to skip those. I did just that in the statsmodels PR even before seeing your comment here... :)

**Comment by jakirkham** [comment]
JFYI NumPy is planning to release 2.0.0 on June 16th

xref: https://github.com/numpy/numpy/issues/24300#issuecomment-2125643448

**Comment by jakirkham** [comment]
More updates:

* NumPy 2.0.0 was released last week
  * Packaged in conda-forge: https://github.com/conda-forge/numpy-feedstock/pull/316
* The RC channel was dropped from the NumPy 2 migration
  * https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6044
  * Any package rebuilt with NumPy 2 so far remains available
* NumPy 2 migration status
  * Nearly 1/3 of packages are rebuilt
  * More than 2/5 have open PRs
  * A bit more than 1/4 are still waiting
  * https://conda-forge.org/status/migration/numpy2

**Comment by jaimergp** [comment]
Do we still need this issue pinned, or even open?

**Comment by jakirkham** [comment]
Nearly half of the feedstocks affected by this migration are in PR or awaiting PR. Also still seeing activity on projects adding NumPy 2 support

Maybe we can leave this as-is for a bit longer to see if we can get more of these through

**Comment by h-vetinari** [comment]
The migration definitely needs more time, and (while independent) I think this issue should stay open, but we can IMO unpin it if there's any other issue that currently deserves a pinned spot

**Comment by jdblischak** [comment]
> Also we are still waiting on a NumPy 2 compatible TensorFlow: https://github.com/tensorflow/tensorflow/issues/67291

TensorFlow [2.18.0](https://github.com/tensorflow/tensorflow/releases/tag/v2.18.0) supports numpy 2

xref: https://github.com/conda-forge/tensorflow-feedstock/pull/408, https://github.com/conda-forge/tensorflow-feedstock/pull/389

**Comment by traversaro** [comment]
Just to track this somewhere global, I have a few feedstocks where the migration to numpy 2 is blocked by https://github.com/conda-forge/pybind11-feedstock/issues/95 (as the workaround for https://github.com/conda-forge/pybind11-feedstock/issues/95 is to force pybind11 to be < 2.12.0) and as the first version of pybind11 that supports numpy 2 is 2.12.0 . A fix has landed upstream in https://github.com/pybind/pybind11/pull/4953 , but we will probably need a new release of pybind11 to be able to use that in conda-forge.
